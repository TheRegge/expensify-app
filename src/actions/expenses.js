import uuid from 'uuid';
import database from '../firebase/firebase';

// Works this way:
// Synchronous action:
// -------------------
// component call action generator
// action generator returns object
// component dispatches object
// redux store changes

// Asychronous action:
// -------------------
// component call action generator
// action generator returns function
// component dispatches function (use redux-thunk)
// function runs (has the ability to
// dispatch other actions and do 
// whatever it wants).

// ADD_EXPENSE ACTION GENERATOR
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

// This below on works because we have
// installed redux-thunk in the 'configureStore.js'
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        // Update firebase data, then dispatch to update the store
        // also, we are returning the call so we can use promises
        // method chainning in the test .then().then(), because when
        // a then returns something, the returned value is available in
        // the next then()... then(return x).then(x)
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
}


// REMOVE_EXPENSE ACTION GENERATOR
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE ACTION GENERATOR
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
// The action actually changing the store
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// The wrapper to make async call to Firebase
// then trigger `setExpenses` wich changes
// the store
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
            const expenses = [];
            // parse data to work with our store:
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            // Dispatch the 'real' action which
            // puts the data in the redux store:
            dispatch(setExpenses(expenses));
        });
    };
};