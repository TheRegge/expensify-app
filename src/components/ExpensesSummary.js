import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
    <div>
    Viewing {props.expenses.length} expense{props.expenses.length > 1 ? 's':''}.
        Total: {
            numeral(props.expensesTotal/100).format('$0,0.00')}
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        expensesTotal: expensesTotal(selectExpenses(state.expenses,state.filters))
    };
};

export default connect(mapStateToProps)(ExpensesSummary);