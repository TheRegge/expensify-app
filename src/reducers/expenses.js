// Expenses Reducer
const expensesReducerDefaultState = [];
export default (state=expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id );
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        // remove all existing expenses in store and 
        // add a new set of expenses passed
        // in the action (the expenses fetched from
        // firebase) So basically, it loads expenses
        // from Firebase db:
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};
