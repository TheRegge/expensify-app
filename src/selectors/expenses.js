import moment from "moment";


// Get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        // type of startDate is not number returns true... meanning if it is undefined (not a number)
        // startDateMatch should return true, because it is not part of filtering anymore...
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
        return 0;
    });
};