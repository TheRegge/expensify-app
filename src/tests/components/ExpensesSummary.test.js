import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={114195} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render singular expense correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} expensesTotal={195} />);
    expect(wrapper.find('div').text()).toBe('Viewing 1 expense. Total: $1.95');
});

test('should render plural expenses correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={114195} />);
    expect(wrapper.find('div').text()).toBe('Viewing 3 expenses. Total: $1,141.95');
});