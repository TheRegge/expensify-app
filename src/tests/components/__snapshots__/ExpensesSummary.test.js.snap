// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  Viewing 
  3
   expense
  s
  . Total: 
  $1,141.95
</div>
`;
